### PR TITLE
[DAG] Fold bitreverse(shl/srl(bitreverse(x),y)) -> srl/shl(x,y)

### PR DESCRIPTION
**Pull request by RKSimon**
Noticed while investigating GFNI per-element vector shifts (we can form SHL but not SRL/SRA)

Alive2: https://alive2.llvm.org/ce/z/fSH-rf

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/bitreverse-shift.ll [hunk 1]
```diff
@@ -1,10 +1,10 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
 ; RUN: llc -mtriple=riscv32 -mattr=+zbkb -verify-machineinstrs < %s \
-; RUN:   | FileCheck %s -check-prefixes=RV32ZBKB
+; RUN:   | FileCheck %s -check-prefixes=CHECK,RV32ZBKB
 ; RUN: llc -mtriple=riscv64 -mattr=+zbkb -verify-machineinstrs < %s \
-; RUN:   | FileCheck %s -check-prefixes=RV64ZBKB
+; RUN:   | FileCheck %s -check-prefixes=CHECK,RV64ZBKB
 
-; TODO: These tests can be optmised
+; These tests can be optmised
```
Comment (jayfoad) [inline llvm/test/CodeGen/RISCV/bitreverse-shift.ll:7]: Typo "optimised" :)

**Review by arsenm — COMMENTED** [review]
Can you also do the globalisel equivalent?

**Comment by goldsteinn** [comment]
LGTM but wait on 1 more.

**Review by arsenm — APPROVED** [review]

**Comment by RKSimon** [comment]
> Can you also do the globalisel equivalent?

Sure, I haven't worked with combinerhelper much, but I'm assuming it shouldn't be that tricky.

**Pull request merged**